File: app/src/main/java/com/ware/widget/recycler/MoreRecyclerView.java
package com.ware.widget.recycler;

import android.content.Context;
import android.util.AttributeSet;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.recyclerview.widget.StaggeredGridLayoutManager;

import org.jetbrains.annotations.NotNull;

import java.util.List;

import static com.ware.widget.recycler.FooterViewKt.hasMore;

public class MoreRecyclerView extends RecyclerView {
    private static final int TYPE_FOOTER = 122321;
    private static final String TAG = "MoreRecyclerView";
    private WrapAdapter mWrapAdapter;
    private LoadListener mLoadListener;
    private boolean mEnableMore = true;

    private final RecyclerView.AdapterDataObserver mDataObserver = new DataObserver();
    private Status mStatus = Status.MORE;

    public MoreRecyclerView(Context context) {
        this(context, null, 0);
    }

    public MoreRecyclerView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public MoreRecyclerView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
    }

    @Override
    @SuppressWarnings("unchecked")
    public void setAdapter(Adapter adapter) {
        mWrapAdapter = new WrapAdapter(adapter);
        super.setAdapter(mWrapAdapter);
        adapter.registerAdapterDataObserver(mDataObserver);
        mDataObserver.onChanged();
    }

    @Override
    public Adapter getAdapter() {
        if (mWrapAdapter != null) {
            return mWrapAdapter.getOriAdapter();
        }

        return null;
    }

    @Override
    public void setLayoutManager(LayoutManager layout) {
        super.setLayoutManager(layout);
        if (mWrapAdapter != null) {
            if (layout instanceof GridLayoutManager) {
                mWrapAdapter.resetGridSpanSize(((GridLayoutManager) layout));
            }
        }
    }

    @Override
    public void onScrollStateChanged(int state) {
        if (state != SCROLL_STATE_IDLE && !hasMore(mStatus)) return;

        LayoutManager layoutManager = getLayoutManager();
        if (layoutManager == null) return;

        if (layoutManager instanceof GridLayoutManager) {
            int mLastVisibleItemPosition = ((GridLayoutManager) layoutManager).findLastCompletelyVisibleItemPosition();
            int itemCount = layoutManager.getItemCount();
            if (mLastVisibleItemPosition == itemCount) {
                updateStatus(Status.LOADING);
                mLoadListener.loadMore();
            }
        }
    }

    public void updateStatus(Status status) {
        mStatus = status;
    }

    private class WrapAdapter extends RecyclerView.Adapter<ViewHolder> {
        private RecyclerView.Adapter<ViewHolder> adapter;

        WrapAdapter(RecyclerView.Adapter<ViewHolder> adapter) {
            this.adapter = adapter;
        }

        RecyclerView.Adapter getOriAdapter() {
            return this.adapter;
        }

        @NotNull
        @Override
        public RecyclerView.ViewHolder onCreateViewHolder(@NotNull ViewGroup parent, int viewType) {
            if (viewType == TYPE_FOOTER) {
                return new FooterHolder(new FooterView(parent.getContext()));
            }
            return adapter.onCreateViewHolder(parent, viewType);
        }

        @Override
        public void onBindViewHolder(@NotNull RecyclerView.ViewHolder holder, int position) {
            boolean footer = isFooter(position);
            if (footer) {
                ((FooterHolder) holder).bindView(mStatus);
            } else if (adapter != null) {
                int itemCount = adapter.getItemCount();
                if (position < itemCount) {
                    adapter.onBindViewHolder(holder, position);
                }
            }
        }

        @Override
        public void onBindViewHolder(@NotNull RecyclerView.ViewHolder holder, int position, @NotNull List<Object> payloads) {
            boolean footer = isFooter(position);
            if (footer) {
                ((FooterHolder) holder).bindView(mStatus);
            } else if (adapter != null) {
                int itemCount = adapter.getItemCount();
                if (position < itemCount) {
                    if (payloads.isEmpty()) {
                        adapter.onBindViewHolder(holder, position);
                    } else {
                        adapter.onBindViewHolder(holder, position, payloads);
                    }
                }
            }
        }

        class FooterHolder extends RecyclerView.ViewHolder {
            public FooterHolder(@NonNull FooterView itemView) {
                super(itemView);
            }

            public void bindView(Status status) {
                ((FooterView) itemView).updateState(status);
            }
        }

        @Override
        public int getItemViewType(int position) {
            if (isFooter(position)) {
                return TYPE_FOOTER;
            }
            if (adapter != null) {
                if (position < adapter.getItemCount()) {
                    int type = adapter.getItemViewType(position);
                    if (viewTypeConflict(type)) {
                        throw new IllegalStateException("XRecyclerView require itemViewType in adapter should be less than 10000 ");
                    }
                    return type;
                }
            }
            return super.getItemViewType(position);
        }

        @Override
        public int getItemCount() {
            int addSize = (mEnableMore ? 1 : 0);
            if (adapter != null) {
                return adapter.getItemCount() + addSize;
            }
            return addSize;
        }

        @Override
        public long getItemId(int position) {
            if (adapter != null && position < adapter.getItemCount()) {
                if (position < adapter.getItemCount()) {
                    return adapter.getItemId(position);
                }
            }
            return super.getItemId(position);
        }

        boolean isFooter(int position) {
            return (mEnableMore && position == getItemCount() - 1);
        }

        private boolean viewTypeConflict(int viewType) {
            return viewType == TYPE_FOOTER;
        }

        private void resetGridSpanSize(GridLayoutManager manager) {
            final GridLayoutManager gridManager = manager;
            GridLayoutManager.SpanSizeLookup oriLookup = gridManager.getSpanSizeLookup();
            GridLayoutManager.SpanSizeLookup newLookup = new GridLayoutManager.SpanSizeLookup() {
                @Override
                public int getSpanSize(int position) {
                    return (isFooter(position)) ? gridManager.getSpanCount() : oriLookup.getSpanSize(position);
                }
            };
            gridManager.setSpanSizeLookup(newLookup);
        }

        @Override
        public void onAttachedToRecyclerView(@NotNull RecyclerView recyclerView) {
            super.onAttachedToRecyclerView(recyclerView);
            RecyclerView.LayoutManager manager = recyclerView.getLayoutManager();
            if (manager instanceof GridLayoutManager) {
                resetGridSpanSize((GridLayoutManager) manager);
            }
            adapter.onAttachedToRecyclerView(recyclerView);
        }

        @Override
        public void onDetachedFromRecyclerView(@NotNull RecyclerView recyclerView) {
            adapter.onDetachedFromRecyclerView(recyclerView);
        }

        @Override
        public void onViewAttachedToWindow(@NotNull RecyclerView.ViewHolder holder) {
            super.onViewAttachedToWindow(holder);
            ViewGroup.LayoutParams lp = holder.itemView.getLayoutParams();
            if (lp instanceof StaggeredGridLayoutManager.LayoutParams && (isFooter(holder.getLayoutPosition()))) {
                StaggeredGridLayoutManager.LayoutParams p = (StaggeredGridLayoutManager.LayoutParams) lp;
                p.setFullSpan(true);
            }
            adapter.onViewAttachedToWindow(holder);
        }

        @Override
        public void onViewDetachedFromWindow(@NotNull RecyclerView.ViewHolder holder) {
            adapter.onViewDetachedFromWindow(holder);
        }

        @Override
        public void onViewRecycled(@NotNull RecyclerView.ViewHolder holder) {
            adapter.onViewRecycled(holder);
        }

        @Override
        public boolean onFailedToRecycleView(@NotNull RecyclerView.ViewHolder holder) {
            return adapter.onFailedToRecycleView(holder);
        }

        @Override
        public void unregisterAdapterDataObserver(@NotNull AdapterDataObserver observer) {
            adapter.unregisterAdapterDataObserver(observer);
        }

        @Override
        public void registerAdapterDataObserver(@NotNull AdapterDataObserver observer) {
            adapter.registerAdapterDataObserver(observer);
        }
    }

    private class DataObserver extends RecyclerView.AdapterDataObserver {
        @Override
        public void onChanged() {
            if (mWrapAdapter != null) {
                mWrapAdapter.notifyDataSetChanged();
            }
        }

        @Override
        public void onItemRangeInserted(int positionStart, int itemCount) {
            mWrapAdapter.notifyItemRangeInserted(positionStart, itemCount);
        }

        @Override
        public void onItemRangeChanged(int positionStart, int itemCount) {
            mWrapAdapter.notifyItemRangeChanged(positionStart, itemCount);
        }

        @Override
        public void onItemRangeChanged(int positionStart, int itemCount, Object payload) {
            mWrapAdapter.notifyItemRangeChanged(positionStart, itemCount, payload);
        }

        @Override
        public void onItemRangeRemoved(int positionStart, int itemCount) {
            mWrapAdapter.notifyItemRangeRemoved(positionStart, itemCount);
        }

        @Override
        public void onItemRangeMoved(int fromPosition, int toPosition, int itemCount) {
            mWrapAdapter.notifyItemMoved(fromPosition, toPosition);
        }
    }

    public void setLoadListener(LoadListener listener) {
        mLoadListener = listener;
    }

    public interface LoadListener {
        void loadMore();
    }
}

File: app/src/main/java/com/ware/widget/recycler/MoreRecyclerView.kt
package com.ware.widget.recycler

import android.content.Context
import android.util.AttributeSet
import android.util.Log
import android.view.ViewGroup
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.GridLayoutManager.SpanSizeLookup
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.StaggeredGridLayoutManager

/**
 * mark:
 * 1. onBindViewHolder payloads
 * 2. registerAdapterDataObserver
 * 3. notifyDataSetChanged() / notifyItemRangeChanged(positionStart, itemCount) 造成的　onCreateViewHolder区别
 */
class MoreRecyclerView @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, defStyle: Int = 0) : RecyclerView(context, attrs, defStyle) {
    private var mWrapAdapter: WrapAdapter? = null
    private var mLoadListener: LoadListener? = null
    private var mEnableMore = true
    private val mDataObserver: AdapterDataObserver = DataObserver()
    private var mStatus = Status.MORE
    override fun setAdapter(adapter: Adapter<in ViewHolder>?) {
        adapter?.let {
            mWrapAdapter = WrapAdapter(it)
            super.setAdapter(mWrapAdapter)

            it.registerAdapterDataObserver(mDataObserver)
            mDataObserver.onChanged()
        }

    }

    override fun getAdapter(): Adapter<*>? {
        return mWrapAdapter?.oriAdapter
    }

    fun setEnableMore(enable: Boolean) {
        mEnableMore = enable
    }

    override fun setLayoutManager(layout: LayoutManager?) {
        super.setLayoutManager(layout)
        if (mWrapAdapter != null) {
            if (layout is GridLayoutManager) {
                mWrapAdapter?.resetGridSpanSize(layout)
            }
        }
    }

    override fun onScrollStateChanged(state: Int) {
        if (state != SCROLL_STATE_IDLE && !hasMore(mStatus)) return
        val layoutManager = layoutManager ?: return
        var lastVisibleItemPosition = NO_POSITION
        val itemCount = layoutManager.itemCount
        when (layoutManager) {
            is GridLayoutManager -> {
                lastVisibleItemPosition = layoutManager.findLastCompletelyVisibleItemPosition()
            }
            is LinearLayoutManager -> {
                lastVisibleItemPosition = layoutManager.findLastCompletelyVisibleItemPosition()
            }
            is StaggeredGridLayoutManager -> {
                val into = IntArray(layoutManager.spanCount)
                layoutManager.findLastCompletelyVisibleItemPositions(into)
                lastVisibleItemPosition = findLast(into)
            }
        }
        if (lastVisibleItemPosition == itemCount) {
            updateStatus(Status.LOADING)
            mLoadListener!!.loadMore()
        }
    }

    private fun findLast(lastPositions: IntArray): Int {
        var max = lastPositions[0]
        for (value in lastPositions) {
            if (value > max) {
                max = value
            }
        }
        return max
    }

    fun updateStatus(status: Status) {
        mStatus = status
    }

    private inner class WrapAdapter internal constructor(private val adapter: Adapter<ViewHolder>) : Adapter<ViewHolder>() {
        val oriAdapter: Adapter<ViewHolder>
            get() = adapter

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
            return if (viewType == TYPE_FOOTER) {
                FooterHolder(FooterView(parent.context))
            } else adapter.onCreateViewHolder(parent, viewType)
        }

        override fun onBindViewHolder(holder: ViewHolder, position: Int) {
            Log.d(TAG, "onBindViewHolder: ")
            if (holder is FooterHolder) {
                holder.bindView(mStatus)
            } else {
                val itemCount = adapter.itemCount
                if (position < itemCount) {
                    adapter.onBindViewHolder(holder, position)
                }
            }
        }

        override fun onBindViewHolder(holder: ViewHolder, position: Int, payloads: List<Any>) {
            Log.d(TAG, "onBindViewHolder: 222222222222222222222222222")
            val footer = isFooter(position)
            if (footer) {
                (holder as FooterHolder).bindView(mStatus)
            } else {
                val itemCount = adapter.itemCount
                if (position < itemCount) {
                    if (payloads.isEmpty()) {
                        adapter.onBindViewHolder(holder, position)
                    } else {
                        adapter.onBindViewHolder(holder, position, payloads)
                    }
                }
            }
        }

        internal inner class FooterHolder(itemView: FooterView) : ViewHolder(itemView) {
            fun bindView(status: Status?) {
                (itemView as FooterView).updateState(status!!)
            }

            init {
                itemView.layoutParams = LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT)
            }
        }

        override fun getItemViewType(position: Int): Int {
            if (isFooter(position)) {
                return TYPE_FOOTER
            }
            if (position < adapter.itemCount) {
                val type = adapter.getItemViewType(position)
                check(!viewTypeConflict(type)) { "itemViewType in adapter conflict with MoreRecyclerView's footer type" }
                return type
            }
            return super.getItemViewType(position)
        }

        override fun getItemCount(): Int {
            val addSize = if (mEnableMore) 1 else 0
            return adapter.itemCount + addSize
        }

        override fun getItemId(position: Int): Long {
            if (position < adapter.itemCount) {
                if (position < adapter.itemCount) {
                    return adapter.getItemId(position)
                }
            }
            return super.getItemId(position)
        }

        fun isFooter(position: Int): Boolean {
            return mEnableMore && position == itemCount - 1
        }

        private fun viewTypeConflict(viewType: Int): Boolean {
            return viewType == TYPE_FOOTER
        }

        fun resetGridSpanSize(manager: GridLayoutManager) {
            val oriLookup = manager.spanSizeLookup
            val newLookup: SpanSizeLookup = object : SpanSizeLookup() {
                override fun getSpanSize(position: Int): Int {
                    return if (isFooter(position)) manager.spanCount else oriLookup.getSpanSize(position)
                }
            }
            manager.spanSizeLookup = newLookup
        }

        override fun onAttachedToRecyclerView(recyclerView: RecyclerView) {
            super.onAttachedToRecyclerView(recyclerView)
            val manager = recyclerView.layoutManager
            if (manager is GridLayoutManager) {
                resetGridSpanSize(manager)
            }
            adapter.onAttachedToRecyclerView(recyclerView)
        }

        override fun onDetachedFromRecyclerView(recyclerView: RecyclerView) {
            adapter.onDetachedFromRecyclerView(recyclerView)
        }

        override fun onViewAttachedToWindow(holder: ViewHolder) {
            super.onViewAttachedToWindow(holder)
            val lp = holder.itemView.layoutParams
            if (lp is StaggeredGridLayoutManager.LayoutParams && isFooter(holder.layoutPosition)) {
                lp.isFullSpan = true
            }
            adapter.onViewAttachedToWindow(holder)
        }

        override fun onViewDetachedFromWindow(holder: ViewHolder) {
            adapter.onViewDetachedFromWindow(holder)
        }

        override fun onViewRecycled(holder: ViewHolder) {
            adapter.onViewRecycled(holder)
        }

        override fun onFailedToRecycleView(holder: ViewHolder): Boolean {
            return adapter.onFailedToRecycleView(holder)
        }

        override fun unregisterAdapterDataObserver(observer: AdapterDataObserver) {
            adapter.unregisterAdapterDataObserver(observer)
        }

        override fun registerAdapterDataObserver(observer: AdapterDataObserver) {
            adapter.registerAdapterDataObserver(observer)
        }

    }

    private inner class DataObserver : AdapterDataObserver() {
        override fun onChanged() {
            mWrapAdapter?.notifyDataSetChanged()
        }

        override fun onItemRangeInserted(positionStart: Int, itemCount: Int) {
            mWrapAdapter?.notifyItemRangeInserted(positionStart, itemCount)
        }

        override fun onItemRangeChanged(positionStart: Int, itemCount: Int) {
            mWrapAdapter?.notifyItemRangeChanged(positionStart, itemCount)
        }

        override fun onItemRangeChanged(positionStart: Int, itemCount: Int, payload: Any?) {
            mWrapAdapter?.notifyItemRangeChanged(positionStart, itemCount, payload)
        }

        override fun onItemRangeRemoved(positionStart: Int, itemCount: Int) {
            mWrapAdapter?.notifyItemRangeRemoved(positionStart, itemCount)
        }

        override fun onItemRangeMoved(fromPosition: Int, toPosition: Int, itemCount: Int) {
            mWrapAdapter?.notifyItemMoved(fromPosition, toPosition)
        }
    }

    fun setLoadListener(listener: LoadListener) {
        mLoadListener = listener
    }

    interface LoadListener {
        fun loadMore()
    }

    companion object {
        private const val TYPE_FOOTER = 122321
        private const val TAG = "MoreRecyclerView"
    }
}